### ISSUE TITLE
How to use borderType AF_PAD_SYM for convolve operations?

### ISSUE THREAD

**Issue by krish021195** [arrayfire/arrayfire#2676]
Hi,

The convolve operations in ArrayFire seems to use the AF_PAD_ZERO borderType. How to change this borderType to AF_PAD_SYM?

Thanks

**Comment by 9prady9** [comment]
Currently, border type can be set only for median, min and max filters only. Convolutions only do zero padding as of now.

**Comment by krish021195** [comment]
Can it be expected in the next release?

For a workaround , Trying to provide a border replicated array to the convolve function. 
My method seems to take more time for just replicating the border.

Is the process of replicating the borders done directly in cuda kernel of median or min or max filters? Is there any simple way of replicating the borders in ArrayFire?

**Comment by 9prady9** [comment]
Unfortunately, we don't have it in the line for the next release.

http://arrayfire.org/docs/indexing.htm You can use indexing feature to create padding with the values of your choice.

Can you please share the code that you used. It would help to see whats exactly is going from code.

**Comment by krish021195** [comment]
I am trying to use indexing feature for padding aswell. But it seems to take more time , the time taken for padding nearly equals the time for convolution operation with the padded array.

I have used the below code for padding,

`
array ar_A = constant(0,1924,1084,3,2,u8);
array ar_B = randu(1920,1080,3,2,u8);

ar_A(seq(2,ar_A.dims(0)-3,1),seq(2,ar_A.dims(1)-3,1)) = ar_B;
ar_A(0,seq(2,ar_A.dims(1)-3,1)) += ar_B(2,span);
ar_A(1,seq(2,ar_A.dims(1)-3,1)) += ar_B(1,span);
ar_A(ar_A.dims(0)-1,seq(2,ar_A.dims(1)-3,1)) += ar_B(ar_B.dims(0)-3,span);
ar_A(ar_A.dims(0)-2,seq(2,ar_A.dims(1)-3,1)) += ar_B(ar_B.dims(0)-2,span);
ar_A(span,0) += ar_A(span,4);
ar_A(span,1) += ar_A(span,3);
ar_A(span,ar_A.dims(1)-1) += ar_A(span,ar_A.dims(1)-5);
ar_A(span,ar_A.dims(1)-2) += ar_A(span,ar_A.dims(1)-4);
`

**Comment by krish021195** [comment]
@9prady9 , Any suggestions on how to improve the timings of the above code?

**Comment by 9prady9** [comment]
Sorry, I haven't looked into it yet, i will get back to you over the
weekend.

**Comment by 9prady9** [comment]
The code seems okay, it wasn't too slow on GTX 1060 GPU - the padding operation took about approximately 2.1 ms on an average for two 3-channel HD images. Can you please share the details of your system, GPU & backend along with the times you are getting on your device.

Note: I used `timeit` (http://arrayfire.org/docs/timing.htm ) to benchmark the code

**Comment by krish021195** [comment]
@9prady9  I am also using timeit to benchmark the code. Using Nvidia Jetson Xavier and backend as cuda. I am getting around ~4ms. Just curious on why the time taken for border replicate is nearly equal to a convolution operation of a 5*5 kernel on two 3-channel HD images.

**Comment by 9prady9** [comment]
I think that is a coincidence. For example, if we keep increasing the kernel size, I doubt padding using the above code will run at same time as convolution as the default zero padding done by convolve operation is much faster compared to looking up values across image border to create mirror padding for larger kernel sizes. Perhaps, a custom kernel would help for larger padding sizes. We have an internal implementation that we haven't exposed yet, I want to benchmark the above code against that before we publish that new padding function.

On a side note: global read/write efficiency won't be 100% for mirror padding because while we are creating the padded columns, we are reading from continuous memory locations hence mem reads would most likely be optimized. However, when we are mirroring the rows, each successive element we are copying to padded area is separated by a distance of `dims[0]` (`strides[1]`)  bringing down read/write efficiency.

I ran a nvprof with metrics `gld_efficiency` and `gst_efficiency` for the code we are discussing now and the results came out as I suspected, the average global load/store efficiency is quite lower than 100% for copy kernel (backend implementation).
```
Device "GeForce GTX 1060 3GB (0)"
    Kernel: void cuda::kernel::copy_kernel<unsigned char, unsigned char, bool=1>
```

```
nvocations     Metric Name               Metric Description     Min      Max      Avg
        135  gld_efficiency    Global Memory Load Efficiency   3.12%  100.00%   36.57%
        135  gst_efficiency   Global Memory Store Efficiency   3.12%   66.07%   30.85%
```

**Comment by krish021195** [comment]
@9prady9 Thanks on the comment. When can we expect that new padding function? Also can you share us the code?

What time does it take when running with two 3-channel HD images on GTX 1060 GPU with cuda backend?

**Comment by 9prady9** [comment]
Hopefully, this week. Open github file finder https://github.com/arrayfire/arrayfire/find/master and search for the term **pad_array_border** that should point to all the files related to array border padding implementation.

As I said earlier, I don't have any benchmarking numbers for this kernel based padding approach yet. There are no numbers on that yet, I will generate them once I create the API. Hopefully, we will have these by end of the week.

**Comment by 9prady9** [comment]
Here are some numbers for the HD image dimensions

```
ArrayFire v3.7.0 (CUDA, 64-bit Linux, build c30d5455f)
Platform: CUDA Toolkit 10.1, Driver: 440.36
[0] GeForce GTX 1060 3GB, 3012 MB, CUDA Compute 6.1
indexing based padding avg runtime: 0.00250179
custom kernel  padding avg runtime: 0.000964645
out dims: 1924 1084 3 2
```

kernel version of padding is giving about **2.6 times** speedup than the indexing based one. Of course this is for symmetric padding, there probably wouldn't be too much of difference for zero padding I would think.

As we increase the padding size, there could be much higher speedup and the code would still look the same where as the indexing based one would transform into some loop based kind of construct.

**Comment by krish021195** [comment]
@9prady9 thanks on sharing the timings. Will keep in track with the issue #2682 .

Can you share me the gpu utilization of the api to pad array borders?

Also Is there anything needs to be done at installation or building steps of ArrayFire to increase the performance ?

**Comment by 9prady9** [comment]
The utilization was at 80% constantly. There is no need to make any changes at environment level to draw maximum performance, that will happen automatically depending on your device and other current loads on the GPU at that moment.

**Comment by 9prady9** [comment]
@krish021195 pad is available now via 3.7.1 release. Closing this issue.
